lib/msgpax/packer.ex

```diff
@@ -236,21 +236,25 @@ defimpl Msgpax.Packer, for: Msgpax.Ext do
 end
 
 defimpl Msgpax.Packer, for: Any do
-  defmacro __deriving__(module, _, opts) do
-    deriving(module, opts)
+  defmacro __deriving__(module, struct, options) do
+    deriving(module, struct, options)
   end
 
-  def deriving(module, opts) do
-    fields = opts[:fields]
+  def deriving(module, struct, options) do
+    keys = struct |> Map.from_struct() |> Map.keys()
+    fields = Keyword.get(options, :fields, keys)
+    include_struct_field? = Keyword.get(options, :include_struct_field, :__struct__ in fields)
+    fields = List.delete(fields, :__struct__)
     extractor =
       cond do
-        fields && :__struct__ in fields ->
-          fields = List.delete(fields, :__struct__)
+        fields == keys and include_struct_field? ->
+          quote(do: Map.from_struct(struct) |> Map.put("__struct__", unquote(module)))
+        fields == keys ->
+          quote(do: Map.from_struct(struct))
+        include_struct_field? ->
```

Comment:
Can't we kill the `fields == keys` clauses and let `Map.take/2` perform basically the same work as `Map.from_struct/1` for the cases when they are the same?